### DISCUSSION TITLE
VSCode cannot find PIL.ImageDraw.ImageDraw "rounded_rectangle" function. Due to bad typings ?

### DISCUSSION THREAD

**Discussion by mat100payette** [python-pillow/Pillow#5512] [Q&A] [answered]
As the title says, whenever I try to use the "rounded_rectangle" function on a ImageDraw object, VSCode cannot directly navigate to it. However, if I navigate to the PIL.ImageDraw.ImageDraw class, I can definitely see that function right after the "rectangle" function.

Is this expected behavior or am I missing something ?

Moreover, on my linux build, it fails even at runtime (I made sure I have >=8.2.0 as a dependency). Can this be related at all ?

**Comment by radarhere** [comment, marked as answer]
What error does it fail with at runtime on Linux?

**Reply by mat100payette** [reply]
Literally just 

> 'ImageDraw' object has no attribute 'rounded_rectangle'

**Reply by radarhere** [reply]
And this doesn't happen if you just try with `rectangle()` instead of `rounded_rectangle()`? My only idea is to suggest that you might have more than one Pillow installation, and to check `Image.__version__` - I recognise this probably isn't all that helpful, since you already say you've made sure you have 8.2.0.

**Reply by mat100payette** [reply]
I'll try that check tomorrow to make extra sure! In my requirements.txt I do have the specific >=8.2.0 version though, so it really should be fine since the linux machine is created from scratch everytime (it's on serverless stuff).

That aside, do you have any idea why vscode can't navigate to that specific function ?

(Also you're correct, `rectangle()` can be navigated to + it works on the linux machine)

**Reply by radarhere** [reply]
No, no ideas apart from checking the version. I can't see any differences between `rounded_rectangle()` and `rectangle()` in the [code](https://github.com/python-pillow/Pillow/blob/22e1fff75453f5bafe6b0016aee5ec679d182592/src/PIL/ImageDraw.py#L252-L341) or the [docs](https://github.com/python-pillow/Pillow/blob/master/docs/reference/ImageDraw.rst) that might cause that.

**Reply by mat100payette** [reply]
Here is the code:

```py
from PIL import Image, ImageDraw
pil_ver = Image.__version__
image: Image.Image = Image.new()
draw: ImageDraw.ImageDraw = ImageDraw.Draw(image)
draw.rectangle()
draw.rounded_rectangle()
```

`pil_ver` shows `'8.2.0'`.

However, [this](https://i.imgur.com/U7sgo16.png) is what I see in VSCode (see linked image).

I really dont understand what's up with that, but I know that Pillow has some pretty nasty type inference and stuff, so maybe it's that... idk.

**Reply by nulano** [reply]
Do all other functions (such as `draw.polygon`) highlight properly?

I don't use VSCode so I have no idea if it will help, but various StackOverflow comments suggest deleting the `.vs` folder can reset the syntax highlighting cache.

**Reply by mat100payette** [reply]
Here is the list of which functions vscode suggest when I do `draw.`:

- arc
- bitmap
- chord
- ellipse
- getfont
- line
- multiline_text
- multiline_textsize
- pieslice
- point
- polygon
- rectangle
- shape
- text
- textsize

As you can see, `rounded_rectangle` is nowhere to be found, despite it being in the `ImageDraw` class.

**Reply by nulano** [reply]
Given that `regular_polygon` ([added in 8.0.0](https://pillow.readthedocs.io/en/stable/releasenotes/8.0.0.html#imagedraw-regular-polygon)) is also not in the list, it looks like your vscode is showing only functions added before version 8.0.0 (i.e. 7.2.0 and earlier, at least for ImageDraw).

One idea is that vscode might be looking at a different Python install. Do you perhaps have Pillow installed in a Python 2.7 install? The last Pillow version to support Python 2.7 is Pillow 6.2.2, which AFAICT also includes that exact list of functions in ImageDraw.

**Reply by mat100payette** [reply]
Well I use Anaconda, and my vscode does show which environment is currently in use (in my case it's `Python 3.8.5 64-bit (my-env-name: conda)`. Also when I navigate to the `ImageDraw` class with ctrl-click, the file path shows that exact environment.

PS I did try to delete `.vs` and restart vscode, but that didn't help.
